controllers/hazelcast/hot_backup_controller.go

```diff
@@ -282,60 +174,151 @@ func (r *HotBackupReconciler) removeSchedule(key types.NamespacedName, logger lo
+	}
+
+	c, ok := GetClient(hazelcastName)
+	if !ok {
+		return r.updateStatus(ctx, backupName, failedHbStatus(errBackupClientNotFound))
+	}
+	c.updateMembers(context.TODO())
+
+	if len(c.Status.MemberMap) == 0 {
+		return r.updateStatus(ctx, backupName, failedHbStatus(errBackupClientNoMembers))
+	}
+
+	g, groupCtx := errgroup.WithContext(ctx)
+
+	// for each member monitor and upload backup if needed
+	for uuid, member := range c.Status.MemberMap {
+		uuid, memberAddress := uuid, member.Address
+		g.Go(func() error {
+			logger := logger.WithValues("uuid", uuid)
+
+			logger.Info("Member status monitor started")
+			defer logger.Info("Member status monitor finished")
+
+			logger.Info("Wait for member backup to finish")
+			if err := waitUntilBackupSucceed(groupCtx, c, uuid); err != nil {
 				return err
 			}
-			return fmt.Errorf("failed to get HotBackup: %w", err)
-		}
-		if hb.Status.State.IsFinished() {
-			if hb.Status.State == hazelcastv1alpha1.HotBackupWaiting {
-				err := agentRest.UploadBackup(ctx)
-				if err != nil {
-					return fmt.Errorf("failed to upload backup folders to external storage: %w", err)
-				}
+
+			// skip upload for local backup
+			if !hz.Spec.Persistence.IsExternal() {
 				return nil
-			} else if hb.Status.State == hazelcastv1alpha1.HotBackupFailure {
-				return errors.New("HotBackup task failed")
 			}
-		} else {
-			logger.Info("HotBackup task is not finished yet. Waiting...")
-			time.Sleep(1 * time.Second)
-		}
+
+			hb := &hazelcastv1alpha1.HotBackup{}
+			if err := r.Get(groupCtx, backupName, hb); err != nil {
+				return err
+			}
+
+			logger.Info("Start and wait for member backup upload")
+			u, err := upload.NewUpload(groupCtx, &upload.Config{
+				MemberAddress: memberAddress,
+				BucketURI:     hb.Spec.BucketURI,
+				BackupPath:    hz.Spec.Persistence.BaseDir,
+				HazelcastName: hb.Spec.HazelcastResourceName,
+				SecretName:    hb.Spec.Secret,
+			})
+			if err != nil {
+				return err
+			}
+
+			// now start and wait for upload
+			if err := u.Start(); err != nil {
+				return err
+			}
+
+			if err := u.Wait(); err != nil {
+				if errors.Is(err, context.Canceled) {
+					// notify agent so we can cleanup if needed
+					logger.Info("Cancel upload")
+					return u.Cancel()
+				}
+				return err
+			}
+
+			// member success
+			return nil
+		})
+	}
+
+	logger.Info("Waiting for members")
+	if err := g.Wait(); err != nil {
+		logger.Error(err, "One or more members failed, returning first error")
+		return r.updateStatus(ctx, backupName, failedHbStatus(err))
+	}
+
+	logger.Info("All members finished with no errors")
+	return r.updateStatus(ctx, backupName, hbWithStatus(hazelcastv1alpha1.HotBackupSuccess))
+}
+
+func (r *HotBackupReconciler) hotBackup(ctx context.Context, hz *hazelcastv1alpha1.Hazelcast) error {
+	client := NewRestClient(hz)
+	if err := client.ChangeState(ctx, Passive); err != nil {
+		return err
 	}
+	defer client.ChangeState(ctx, Active)
```

Comment:
Old implementation seems okay to solve lint problem. WDYT?
```suggestion
	defer func(client *RestClient) {
		e := client.ChangeState(ctx, Active)
		if e != nil {
			logger.Error(e, "Could not change the cluster state to ACTIVE")
		}
	}(client)
```